src/asmcnc/keyboard/custom_keyboard.py

```diff
@@ -0,0 +1,107 @@
+        super(Keyboard, self).__init__(**kwargs)
+
+        # For python 2 encoding issues
+        reload(sys)
+        sys.setdefaultencoding('utf-8')
+
+        self.text_instance = None
+
+        self.kr_layout = "./asmcnc/keyboard/layouts/kr.json"
+        self.numeric_layout = "./asmcnc/keyboard/layouts/numeric.json"
+        self.qwerty_layout = "./asmcnc/keyboard/layouts/qwerty.json"
+        self.qwertyKR_layout = "./asmcnc/keyboard/layouts/qwertyKR.json"
+
+        self.layout = self.kr_layout
+
+        self.font = './asmcnc/keyboard/fonts/KRFont.ttf'
+
+        self.font_name = self.font
+        self.do_translation = False
+        self.width = Window.width
+        self.height = 250
+        self.pos = (Window.width - self.width, 0)
+        self.on_key_up = self.key_up
+
+        for text_input in text_inputs:
+            text_input.keyboard_mode = 'managed'
+            text_input.bind(focus=self.on_focus)
+            text_input.font_name = self.font
+            text_input.multiline = False
+
+    def setup_text_inputs(self, text_inputs):
+
+
+        for text_input in text_inputs:
+            text_input.keyboard_mode = 'managed'
+            text_input.bind(focus=self.on_focus)
+            text_input.font_name = self.font
+            text_input.multiline = False
+
+    def key_up(self, keycode, internal, modifiers):
+
+        try:
+            if "kr" in self.layout.lower():
+                if keycode == "spacebar":
+                    if hgtk.text.compose(self.text_instance.text + "ᴥ") != self.text_instance.text:
+                        self.text_instance.text = hgtk.text.compose(self.text_instance.text + "ᴥ")
+                        return
+        except:
+            pass
+
+        if self.text_instance:
+            if internal == None:
+                if keycode == "enter":
+                    self.text_instance.text = self.text_instance.text + "\n"
+                if keycode == "Han/Yeong":
+                    #https://en.wikipedia.org/wiki/Language_input_keys#Keys_for_Korean_Keyboards
+                    self.layout = self.qwertyKR_layout if self.layout == self.kr_layout else self.kr_layout
+
+                if keycode == "escape":
+                    self.text_instance.focus = False
+                if keycode == "backspace":
+                    self.text_instance.text = self.text_instance.text[:-1]
+                if keycode == "spacebar":
+                    print("SPace")
+                    self.text_instance.text = self.text_instance.text + " "
+                return
+
+            self.text_instance.text = self.text_instance.text + internal
+
+    def on_focus(self,instance,value):
+        if value:
+            try:
+                instance.get_focus_previous().focus = False
+            except:
+                pass
+            instance.focus = True
+            print(str(instance.hint_text) + " focused")
+            try:
+                Window.add_widget(self)
```

Comment:
why does it have to add and remove the text input widget (if I've understood correctly)?

Reply:
It adds or removes the keyboard widget. Because this class inherits from VKeyboard, self refers to the keyboard itself, not the text input